File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    const visited = {[begin]: 0};
    const queue = [begin];
    
    function isConnected(str1, str2) {
        let count = 0;
           for(let i=0; i<str1.length; i++) {
            if (str1[i] !== str2[i]) count++;
        }
        
        return count === 1 ? true: false;
    }
    
    while(queue.length) {
        let cur = queue.shift();
        
        if (cur === target) break;
        for (const word of words) {
            if (isConnected(word, cur) && !visited[word]) {
                visited[word] = visited[cur] + 1;
                queue.push(word)
            }
        }
    }
    return visited[target] ?? 0
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def can_convert(w1, w2):
    diff = 0
    for a, b in zip(w1, w2):
        if a != b:
            diff += 1
        if diff > 1:
            return False
    return diff == 1

def solution(begin, target, words):
    if target not in words:
        return 0
    
    visited = set()
    q = deque([(begin, 0)])
    
    while q:
        current, depth = q.popleft()
        
        if current == target:
            return depth
        
        for word in words:
            if word not in visited and can_convert(current, word):
                visited.add(word)
                q.append((word, depth+1))
    return 0